File: 2021/day2.go
package main

import (
	"fmt"
	"io"
	"log"
	"os"
)

func main() {
	loc, depth := 0, 0
	aim := 0
	/* part 1
	d := map[string]func(int){
		"forward": func(n int) { loc += n },
		"up":      func(n int) { depth -= n },
		"down":    func(n int) { depth += n },
	}
	*/
	d := map[string]func(int){
		"forward": func(n int) {
			loc += n
			depth += aim * n
		},
		"up":   func(n int) { aim -= n },
		"down": func(n int) { aim += n },
	}

	// file, err := os.Open("2_small.txt")
	file, err := os.Open("2.txt")
	if err != nil {
		log.Fatal(err)
	}
	defer file.Close()

	for {
		var cmd string
		var n int
		_, err := fmt.Fscanf(file, "%s %d", &cmd, &n)
		if err == io.EOF {
			break
		}
		if err != nil {
			log.Fatal(err)
		}
		d[cmd](n)
	}
	fmt.Println(loc * depth)
}

File: 2021/day2.py


# file_name = '2_small.txt'
file_name = '2.txt'

# part 1
if 0:
    loc, depth = 0, 0
    with open(file_name) as fp:
        for line in fp:
            cmd, n = line.split()
            n = int(n)
            if cmd == 'forward':
                loc += n
            elif cmd == 'up':
                depth -= n
            elif cmd == 'down':
                depth += n

aim, depth, loc = 0, 0, 0
with open(file_name) as fp:
    for line in fp:
        cmd, n = line.split()
        n = int(n)
        if cmd == 'down':
            aim += n
        elif cmd == 'up':
            aim -= n
        elif cmd == 'forward':
            loc += n
            depth += aim * n


print(loc * depth)
